iree/compiler/Dialect/HAL/Target/LLVM/AOT/LLVMAOTTarget.cpp

```diff
@@ -66,60 +66,98 @@ class LLVMAOTTargetBackend final : public LLVMBaseTargetBackend {
       return failure();
     }
 
+    // Export all entry points such that they are accessible on the dynamic
+    // libraries we generate.
     iree::DyLibExecutableDefT dyLibExecutableDef;
-    // Create invocation function an populate entry_points.
-    auto entryPointOps = targetOp.getBlock().getOps<ExecutableEntryPointOp>();
-
-    for (auto entryPointOp : entryPointOps) {
+    SmallVector<StringRef, 8> entryPointNames;
+    for (auto entryPointOp :
+         targetOp.getBlock().getOps<ExecutableEntryPointOp>()) {
       dyLibExecutableDef.entry_points.push_back(
           std::string(entryPointOp.sym_name()));
+      entryPointNames.push_back(entryPointOp.sym_name());
     }
 
-    // LLVMIR opt passes.
-    auto targetMachine = createTargetMachine(options_);
-    if (!targetMachine) {
-      targetOp.emitError("Can't create target machine for target triple: " +
-                         options_.targetTriple);
+    // Try to grab a linker tool based on the options (and target environment).
+    llvm::Triple targetTriple(options_.targetTriple);
+    auto linkerTool = LinkerTool::getForTarget(targetTriple, options_);
+    if (!linkerTool) {
       return failure();
     }
 
+    // Configure the module with any code generation options required later by
+    // linking (such as initializer functions).
+    if (failed(
+            linkerTool->configureModule(llvmModule.get(), entryPointNames))) {
+      return targetOp.emitError()
+             << "failed to configure LLVM module for target linker";
+    }
+
+    // LLVM opt passes that perform code generation optimizations/transformation
+    // similar to what a frontend would do before passing to linking.
+    auto targetMachine = createTargetMachine(options_);
+    if (!targetMachine) {
+      return targetOp.emitError()
+             << "Can't create target machine for target triple: "
+             << options_.targetTriple;
```

Comment:
General comment on this sort of error logging: I'd find it more useful to wrap the string in `<>` or `""` to make it clear when it is empty string. Otherwise we can get logs like

> Can't link executable and generate target dylib, using linker toolchain

vs

> Can't link executable and generate target dylib, using linker toolchain ""